Improve pagination

KidkArolis opened the pull request:
Wondering if instead of `defaultPageSize` and `defaultPageSizeWhenFetchingAll` a more general hook for decorating query could be used.. but that is also available as a feathers hook on the client already.. but you need to know if allPages param is set or not, so figbird hook could have the rest of the params passed in to make the decision of how to decorate. And so having said all that `defaultPageSizeWhenFetchingAll={250}` prop kinda feels good nough.

KidkArolis committed:
Upgrade dependencies

KidkArolis committed:
Introduce support for total-less endpoints, and parallelLimit option

KidkArolis committed:
Introduce defaultPageSize and defaultPageSizeWhenFetchingAll

tjenkinson commented on lib/fetch.js:
```diff
@@ -95,3 +121,9 @@ function finder(service, params, { queryId, allPages, parallel }) {
+
+// allow total to be undefined or -1 to indicate
+// that total will not be available on this endpoint
+function isTotalAvailale(res) {
```
```suggestion
function isTotalAvailable(res) {
```

tjenkinson commented on lib/fetch.js:
```diff
@@ -54,7 +71,20 @@ function finder(service, params, { queryId, allPages, parallel }) {
+      //    we optimised a bit and we keep fetching more
+      //  - if all or some parallel requests return blank - it's ok
+      //    we accept the trade off of trying to paralellise
+      const requiredFetches = isTotalAvailale(result)
```
```suggestion
      const requiredFetches = isTotalAvailable(result)
```

tjenkinson commented on lib/fetch.js:
```diff
@@ -45,7 +58,11 @@ function finder(service, params, { queryId, allPages, parallel }) {
+      if (
+        res.data.length === 0 ||
+        res.data.length < res.limit ||
+        (isTotalAvailale(res) && result.data.length >= res.total)
```
```suggestion
        (isTotalAvailable(res) && result.data.length >= res.total)
```

tjenkinson commented on CHANGELOG.md:
```diff
@@ -1,5 +1,12 @@
 
+## 0.17.0
+
+- Introduce support for endpoints that do not respond with `result.total` or `result.total` as -1 when using `allPages: true` to indicate that the endpoint is not counting the total available results, instead figbird will keep paging over the API until all data is gathered. Use `optimisticParallelLimit` to control how many requests to trigger when using in conjunction with `parallel`.
```
Wondering if `optimisticParallelLimit` is actually useful? Would we really want to set a different parallel limit if we don't know how many total pages there are?

KidkArolis replied:
My thinking was roughly that:

* if we know the total number of pages, it's good to go for a bit more, e.g. 4 (right now we fetch them all at once, but I think that's also bad, anything without limits is bad..)
* if we don't know the total number of pages, we can either fetch one at a time, or optimistically launch some requests ahead of time risking they'll come back blank, but in this case maybe 4 is too much, so it's 2
* by separating those 2 settings, we can handle both kind of endpoints optimally in the same application

KidkArolis replied:
* and even if the application only has one kind of endpoints (e.g. with total only, or without total only), then the defaults are better for both cases without you having to configure anything

tjenkinson replied:
sure, sounds good 👍

tjenkinson reviewed (approved):
LGTM! 🚀 Just the question on `optimisticParallelLimit` and also a typo

KidkArolis committed:
Simplify again, remove optimisticParallelLimit

It's a neat idea, but getting a bit complex! In particular, if you
really want to fetch more.. just serve larger page sizes from the server!

KidkArolis committed:
Upgrade deps

KidkArolis committed:
Update docs commands

KidkArolis committed:
Fix the comment
